### PR TITLE
Fix TF-M flash region configuration for nRF54L15

### PR DESCRIPTION
**Pull request by Vge0rge**
See commit messages for descriptions.

### CODE REVIEW

File: doc/nrf/security/tfm.rst [hunk 1]
```diff
@@ -174,22 +174,24 @@ The following limitations apply to TF-M and its usage:
...
-Both the partition start address and the partition size need to be aligned with the NRF_SPU flash region size :kconfig:option:`CONFIG_NRF_SPU_FLASH_REGION_SIZE`.
+The partitions which need to be aligned with the TrustZone flash region size are partitions ``tfm_nonsecure`` and ``nonsecure_storage``.
+Both the partition start address and the partition size need to be aligned with the flash region size :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE`.
 
 Note that the ``tfm_nonsecure`` partition is placed after the ``tfm_secure`` partition, thus the end address of the ``tfm_secure`` partition is the same as the start address of the ``tfm_nonsecure`` partition.
 As a result, altering the size of the ``tfm_secure`` partition affects the start address of the ``tfm_nonsecure`` partition.
 
-The following static partition snippet shows a non-aligned configuration for nRF5340 which has a SPU flash region size :kconfig:option:`CONFIG_NRF_SPU_FLASH_REGION_SIZE` of 0x4000.
+The following static partition snippet shows a non-aligned configuration for nRF5340 which has a TrustZon flash region size :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE` of 0x4000.
```
Comment (SebastianBoe) [inline doc/nrf/security/tfm.rst:194]: typo.

TrustZon

**Review by SebastianBoe — APPROVED** [review]

**Review by vili-nordic — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Review by mswarowsky — APPROVED** [review]

File: doc/nrf/security/tfm.rst [hunk 2]
```diff
@@ -174,22 +174,24 @@ The following limitations apply to TF-M and its usage:
 TF-M partition alignment requirements
 *************************************
 
-TF-M requires that secure and non-secure partition addresses must be aligned to the NRF_SPU flash region size :kconfig:option:`CONFIG_NRF_SPU_FLASH_REGION_SIZE`.
+TF-M requires that secure and non-secure partition addresses must be aligned to the flash region size :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE`.
 |NCS| ensures that they in fact are aligned and comply with the TF-M requirements.
 
-TF-M requires this alignment because it uses the SPU to enforce the security policy between the partitions.
+In nRF53 and nRF91 series TF-M uses the SPU to enforce the security policy between the partitions, so the :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE` is set to the SPU flash region size.
+In nRF54L15 TF-M uses the MPC to enforce the security policy between the partitions, so the :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE` is set to the MPC region size.
```
Comment (mia-ko) [inline doc/nrf/security/tfm.rst:181]:
```suggestion
On nRF53 and nRF91 Series devices, TF-M uses the System Protection Unit (SPU) peripheral to enforce the security policy between the partitions, so the :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE` is set to the SPU flash region size.
On nRF54L15 devices, TF-M uses the Memory Protection Controller (MPC) peripheral to enforce the security policy between the partitions, so the :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE` is set to the MPC region size.
```

File: doc/nrf/security/tfm.rst [hunk 3]
```diff
@@ -174,22 +174,24 @@ The following limitations apply to TF-M and its usage:
...
+
 When the :ref:`partition_manager` is enabled, it will take into consideration the alignment requirements.
 But when the static partitions are used, the user is responsible for following the alignment requirements.
 
 If you are experiencing any partition alignment issues when using the Partition Manager, check the :ref:`known_issues` page on the main branch.
 
-The partitions which need to be aligned to the SPU flash region size are partitions ``tfm_nonsecure`` and ``nonsecure_storage``.
-Both the partition start address and the partition size need to be aligned with the NRF_SPU flash region size :kconfig:option:`CONFIG_NRF_SPU_FLASH_REGION_SIZE`.
+The partitions which need to be aligned with the TrustZone flash region size are partitions ``tfm_nonsecure`` and ``nonsecure_storage``.
+Both the partition start address and the partition size need to be aligned with the flash region size :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE`.
```
Comment (mia-ko) [inline doc/nrf/security/tfm.rst:189]:
```suggestion
The partitions ``tfm_nonsecure`` and ``nonsecure_storage`` need to be aligned with the TrustZone flash region size.
Both the partition start address and the partition size need to be aligned with the flash region size :kconfig:option:`CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE`.
```

File: subsys/partition_manager/pm.yml.zboss [hunk 4]
```diff
@@ -15,7 +15,7 @@ zboss_product_config:
   placement:
     after: [zboss_nvram]
 #ifdef CONFIG_BUILD_WITH_TFM
-    align: {start: CONFIG_NRF_SPU_FLASH_REGION_SIZE}
+    align: {start: CONFIG_NRF_TRUSTZONE_FLASH_REGION_ALIGNMENT}
```
Comment (mia-ko) [inline subsys/partition_manager/pm.yml.zboss:18]: With _SIZE replacing the _ALIGNMENT in the end?

```suggestion
    align: {start: CONFIG_NRF_TRUSTZONE_FLASH_REGION_ALIGNMENT}
```

File: modules/trusted-firmware-m/tfm_boards/common/assert.c [hunk 5]
```diff
@@ -12,21 +12,28 @@
...
+#error "TF-M non-secure start address is not aligned to SPU/MPC HW requirements"
 #endif
 
-#if !(IS_ALIGNED_POW2(PM_SRAM_NONSECURE_ADDRESS, CONFIG_NRF_SPU_RAM_REGION_ALIGNMENT))
-#error "SRAM non-secure address is not aligned to SPU HW requirements"
+#if !(IS_ALIGNED_POW2(PM_SRAM_NONSECURE_ADDRESS, CONFIG_NRF_TRUSTZONE_RAM_REGION_SIZE))
+#pragma message \
+	"SRAM non-secure address is not aligned to SPU/MPC HW requirements" \
+	"\nIn nRF53/nRF91 series the RAM region need to be aligned with the SPU region size." \
+	"\nIn nRF54L15 the RAM region need to be aligned with the MPC region size.\n\n"
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/common/assert.c:36]:
```suggestion
	"SRAM non-secure address is not aligned with SPU/MPC HW requirements" \
	"\nOn nRF53/nRF91 Series devices, the RAM region needs to be aligned with the SPU region size." \
	"\nOn nRF54L15 devices, the RAM region needs to be aligned with the MPC region size.\n\n"
```

File: modules/trusted-firmware-m/tfm_boards/common/assert.c [hunk 6]
```diff
@@ -12,21 +12,28 @@
...
+
+#error "SRAM non-secure start address is not aligned to SPU/MPC HW requirements"
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/common/assert.c:38]:
```suggestion
#error "SRAM non-secure start address is not aligned with SPU/MPC HW requirements"
```

File: modules/trusted-firmware-m/tfm_boards/partition/region_defs.h [hunk 7]
```diff
@@ -58,6 +54,14 @@
 #endif
 #endif
 
+
+/* In nR53 and nRF91 series the SPU is used to configure S/NS/NSC regions. */
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/partition/region_defs.h:58]:
```suggestion
/* On nR53 and nRF91 Series devices, the SPU is used to configure S/NS/NSC regions. */
```

File: modules/trusted-firmware-m/tfm_boards/common/assert.c [hunk 8]
```diff
@@ -12,21 +12,28 @@
...
+#if !(IS_ALIGNED_POW2(PM_TFM_NONSECURE_ADDRESS, CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE))
 #pragma message \
 	"\n\n!!!Partition alignment error!!!" \
 	"\nThe non-secure start address in pm_static.yml" \
 	" or generated partition.yml is: " \
 	M2S(PM_TFM_NONSECURE_ADDRESS) \
-	"\nwhich is not aligned with the SPU HW requirements."				\
+	"\nwhich is not aligned with the SPU/MPC HW requirements."				\
+	"\nIn nRF53/nRF91 series the flash region need to be aligned with the SPU region size." \
+	"\nIn nRF54L15 the flash region need to be aligned with the MPC region size." \
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/common/assert.c:25]:
```suggestion
	"\nOn nRF53/nRF91 Series devices, the flash region needs to be aligned with the SPU region size." \
	"\nOn nRF54L15 devices, the flash region needs to be aligned with the MPC region size." \
```

File: modules/trusted-firmware-m/tfm_boards/common/assert.c [hunk 9]
```diff
@@ -12,21 +12,28 @@
...
 	"\nRefer to the documentation section 'TF-M partition alignment requirements'" \
 	"\nfor more information.\n\n"
 
-#error "TF-M non-secure start address is not aligned to SPU HW requirements"
+#error "TF-M non-secure start address is not aligned to SPU/MPC HW requirements"
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/common/assert.c:29]:
```suggestion
#error "TF-M non-secure start address is not aligned with SPU/MPC HW requirements"
```

File: modules/trusted-firmware-m/tfm_boards/partition/region_defs.h [hunk 10]
```diff
@@ -58,6 +54,14 @@
...
+#if defined(CONFIG_CPU_HAS_NRF_IDAU)
+
+/* These definitions are being used by the spu.c file in the TF-M repo. */
+#define SPU_FLASH_REGION_SIZE     (CONFIG_NRF_TRUSTZONE_FLASH_REGION_SIZE)
+#define SPU_SRAM_REGION_SIZE      (CONFIG_NRF_TRUSTZONE_RAM_REGION_SIZE)
+
 /* The veneers needs to be placed at the end of the secure image.
```
Comment (mia-ko) [inline modules/trusted-firmware-m/tfm_boards/partition/region_defs.h:65]:
```suggestion
/* The veneers need to be placed at the end of the secure image.
```

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Review by annwoj — APPROVED** [review]

File: subsys/partition_manager/pm.yml.zboss (hunk 4 above)
Comment (Vge0rge) [reply subsys/partition_manager/pm.yml.zboss]: Good catch, thanks!

**Review by mia-ko — APPROVED** [review]

**Review by mia-ko — COMMENTED** [review]
Btw, do we need a changelog entry for this?

**Comment by Vge0rge** [comment]
> Btw, do we need a changelog entry for this?

No I would say so, it's basically a restructuring of existing functionality.

**Comment by nordic-piks** [comment]
Please do not postpone merging this PR (even do it faster is possible) as now all PR which are trying to update sdk-zephyr are failing.

**Pull request merged**